subprojects/hashing/src/main/java/org/gradle/internal/hash/HashCode.java

```diff
@@ -95,98 +109,229 @@ private static int decode(char ch) {
+            hasher.putLong(bits1);
+            hasher.putLong(bits2);
+        }
+
+        @Override
+        public int hashCode() {
+            return (int) bits1;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || o.getClass() != HashCode128.class) {
                 return false;
             }
+
+            HashCode128 other = (HashCode128) o;
+
+            return bits1 == other.bits1 && bits2 == other.bits2;
         }
 
-        return true;
+        @Override
+        public int compareTo(HashCode o) {
+            if (o.getClass() != HashCode128.class) {
+                return HashCode.compareBytes(bytes(), o.bytes());
+            }
+
+            HashCode128 other = (HashCode128) o;
+
+            int result = compareLong(bits1, other.bits1);
+            if (result == 0) {
+                result = compareLong(bits2, other.bits2);
+            }
+
+            return result;
+        }
     }
 
-    @Override
-    public int compareTo(@Nonnull HashCode o) {
-        byte[] bytes2 = o.bytes;
+    static class ByteArrayBackedHashCode extends HashCode {
+        private final byte[] bytes;
+        private int hashCode;
+
+        public ByteArrayBackedHashCode(byte[] bytes) {
+            this.bytes = bytes;
+        }
+
+        @Override
+        public int length() {
+            return bytes.length;
+        }
+
+        @Override
+        byte[] bytes() {
+            return bytes;
+        }
+
+        @Override
+        public byte[] toByteArray() {
+            return bytes.clone();
+        }
+
+        @Override
+        void appendToHasher(PrimitiveHasher hasher) {
+            hasher.putBytes(bytes);
+        }
+
+        @Override
+        public int hashCode() {
+            if (hashCode == 0) {
+                hashCode = (int) ((bytes[0] & 0xFF)
+                    | ((bytes[1] & 0xFF) << 8)
+                    | ((bytes[2] & 0xFF) << 16)
+                    | ((bytes[3] & 0xFF) << 24)
+                    // Make sure it's always > 0 but without affecting the lower 32 bits
+                    | (1L << 32));
+            }
+            return hashCode;
+        }
+
+        @Override
+        public boolean equals(@Nullable Object obj) {
+            if (obj == this) {
+                return true;
+            }
+
+            if (obj == null || obj.getClass() != ByteArrayBackedHashCode.class) {
+                return false;
+            }
+
+            byte[] a = bytes;
+            byte[] b = ((ByteArrayBackedHashCode) obj).bytes;
+            int length = a.length;
+
+            if (b.length != length) {
+                return false;
+            }
+
+            for (int i = 0; i < length; i++) {
+                if (a[i] != b[i]) {
+                    return false;
+                }
+            }
+
+            return true;
```

Comment:
```suggestion
            return Arrays.equals(bytes, ((ByteArrayBackedHashCode) obj).bytes);
```

Reply:
I did this already the other way around once, see https://github.com/gradle/gradle/commit/09ad5fbcc7eff421662e4dd17c39e655f0659494. I believe we've measured some performance impact here.

Reply:
I see. A comment would be great as the code is begging for replacing with Arrays :)